pkg/kubectl/rolling_updater.go

```diff
@@ -308,14 +308,14 @@ func (r *RollingUpdater) scaleDown(newRc, oldRc *api.ReplicationController, desi
 		return oldRc, nil
 	}
 	// Block until there are any pods ready.
-	_, newAvailable, err := r.waitForReadyPods(config.Interval, config.Timeout, oldRc, newRc)
+	oldAvailable, newAvailable, err := r.waitForReadyPods(config.Interval, config.Timeout, oldRc, newRc)
```

Comment:
You are right on this scenario. In any case, scaling down unhealthy pods shouldn't affect minAvailable (see https://github.com/kubernetes/kubernetes/pull/20368). I have updated the PR to do so.

The difference between the controller and the rolling updater is that in the former case, pods are constantly being monitored by the controller and any discrepancies will be automatically reconciled (there is still a possibility for things to turn hairy), while in the later case the rolling updater may run against an already violated deployment.